File: lib/music/instruments/tr8.ts
import { createController } from './createController';
import { mapToNote } from './../elements/mapToNote';
import { MidiValue } from '../../midi/MidiValue';

type DrumsetName = 'tr808' | 'tr909';

const drumsets: Record<DrumsetName, MidiValue> = {
  tr808: 0,
  tr909: 1
};

const tr8 = {
  drumset (drumsetName: DrumsetName): MidiValue {
    return drumsets[drumsetName];
  },

  bassDrum: mapToNote('c2'),
  bassDrumTune: createController({ id: 20 }),
  bassDrumAttack: createController({ id: 21 }),
  bassDrumDecay: createController({ id: 23 }),
  bassDrumCompressor: createController({ id: 22 }),

  snareDrum: mapToNote('d2'),
  snareDrumTune: createController({ id: 25 }),
  snareDrumSnappiness: createController({ id: 26 }),
  snareDrumDecay: createController({ id: 28 }),
  snareDrumCompressor: createController({ id: 27 }),

  lowTom: mapToNote('g2'),
  lowTomTune: createController({ id: 46 }),
  lowTomDecay: createController({ id: 47 }),

  midTom: mapToNote('b2'),
  midTomTune: createController({ id: 49 }),
  midTomDecay: createController({ id: 50 }),

  highTom: mapToNote('d3'),
  highTomTune: createController({ id: 52 }),
  highTomDecay: createController({ id: 53 }),

  rimShot: mapToNote('c#2'),
  rimShotTune: createController({ id: 55 }),
  rimShotDecay: createController({ id: 56 }),

  handClap: mapToNote('d#2'),
  handClapTune: createController({ id: 58 }),
  handClapDecay: createController({ id: 59 }),

  closedHihat: mapToNote('f#2'),
  closedHihatTune: createController({ id: 61 }),
  closedHihatDecay: createController({ id: 62 }),

  openHihat: mapToNote('a#2'),
  openHihatTune: createController({ id: 80 }),
  openHihatDecay: createController({ id: 81 }),

  crashCymbal: mapToNote('c#3'),
  crashCymbalTune: createController({ id: 83 }),
  crashCymbalDecay: createController({ id: 84 }),

  rideCymbal: mapToNote('d#3'),
  rideCymbalTune: createController({ id: 89 }),
  rideCymbalDecay: createController({ id: 87 })
};

export { tr8 };

File: synthesizer/tr8.go
package synthesizer

import "gitlab.com/gomidi/midi/v2"

type TR8 struct {
	Synthesizer
}

func NewTR8(synthesizer Generic, err error) (TR8, error) {
	if err != nil {
		return TR8{}, err
	}

	return TR8{synthesizer}, nil
}

func (tr8 TR8) BassDrum() midi.Note {
	return midi.Note(midi.C(3))
}

func (tr8 TR8) BassDrumTune() uint8 {
	return 20
}

func (tr8 TR8) BassDrumAttack() uint8 {
	return 21
}

func (tr8 TR8) BassDrumDecay() uint8 {
	return 23
}

func (tr8 TR8) BassDrumCompressor() uint8 {
	return 22
}

func (tr8 TR8) SnareDrum() midi.Note {
	return midi.Note(midi.D(3))
}

func (tr8 TR8) SnareDrumTune() uint8 {
	return 25
}

func (tr8 TR8) SnareDrumSnappiness() uint8 {
	return 26
}

func (tr8 TR8) SnareDrumDecay() uint8 {
	return 28
}

func (tr8 TR8) SnareDrumCompressor() uint8 {
	return 27
}

func (tr8 TR8) LowTom() midi.Note {
	return midi.Note(midi.G(3))
}

func (tr8 TR8) LowTomTune() uint8 {
	return 46
}

func (tr8 TR8) LowTomDecay() uint8 {
	return 47
}

func (tr8 TR8) MidTom() midi.Note {
	return midi.Note(midi.B(3))
}

func (tr8 TR8) MidTomTune() uint8 {
	return 49
}

func (tr8 TR8) MidTomDecay() uint8 {
	return 50
}

func (tr8 TR8) HighTom() midi.Note {
	return midi.Note(midi.D(4))
}

func (tr8 TR8) HighTomTune() uint8 {
	return 52
}

func (tr8 TR8) HighTomDecay() uint8 {
	return 53
}

func (tr8 TR8) Rimshot() midi.Note {
	return midi.Note(midi.Db(3))
}

func (tr8 TR8) RimshotTune() uint8 {
	return 55
}

func (tr8 TR8) RimshotDecay() uint8 {
	return 56
}

func (tr8 TR8) Handclap() midi.Note {
	return midi.Note(midi.Eb(3))
}

func (tr8 TR8) HandclapTune() uint8 {
	return 58
}

func (tr8 TR8) HandclapDecay() uint8 {
	return 59
}

func (tr8 TR8) ClosedHiHat() midi.Note {
	return midi.Note(midi.Gb(3))
}

func (tr8 TR8) ClosedHiHatTune() uint8 {
	return 61
}

func (tr8 TR8) ClosedHiHatDecay() uint8 {
	return 62
}

func (tr8 TR8) OpenHiHat() midi.Note {
	return midi.Note(midi.Bb(3))
}

func (tr8 TR8) OpenHiHatTune() uint8 {
	return 80
}

func (tr8 TR8) OpenHiHatDecay() uint8 {
	return 81
}

func (tr8 TR8) CrashCymbal() midi.Note {
	return midi.Note(midi.Db(4))
}

func (tr8 TR8) CrashCymbalTune() uint8 {
	return 83
}

func (tr8 TR8) CrashCymbalDecay() uint8 {
	return 84
}

func (tr8 TR8) RideCymbal() midi.Note {
	return midi.Note(midi.Eb(4))
}

func (tr8 TR8) RideCymbalTune() uint8 {
	return 89
}

func (tr8 TR8) RideCymbalDecay() uint8 {
	return 87
}
